// gojs-react/lib/esm/gojsreact.js
// Copyright (C) 1998-2020 by Northwoods Software Corporation. All Rights Reserved.
import { __extends } from 'tslib';
import { Diagram, GraphLinksModel, Overview, Spot } from 'gojs';
import { createRef, createElement, Component } from 'react';

var ReactDiagram = (function (_super) {
    __extends(ReactDiagram, _super);
    function ReactDiagram(props) {
        var _this = _super.call(this, props) || this;
        _this.modelChangedListener = null;
        _this.divRef = createRef();
        return _this;
    }
    ReactDiagram.prototype.getDiagram = function () {
        if (this.divRef.current === null)
            return null;
        return Diagram.fromDiv(this.divRef.current);
    };
    ReactDiagram.prototype.componentDidMount = function () {
        var _this = this;
        if (this.divRef.current === null)
            return;
        var diagram = this.props.initDiagram();
        diagram.div = this.divRef.current;
        this.modelChangedListener = function (e) {
            if (e.isTransactionFinished && e.model && !e.model.isReadOnly && _this.props.onModelChange) {
                var dataChanges = e.model.toIncrementalData(e);
                if (dataChanges !== null)
                    _this.props.onModelChange(dataChanges);
            }
        };
        diagram.addModelChangedListener(this.modelChangedListener);
        diagram.delayInitialization(function () {
            var model = diagram.model;
            model.commit(function (m) {
                m.mergeNodeDataArray(_this.props.nodeDataArray);
                if (_this.props.linkDataArray !== undefined && m instanceof GraphLinksModel) {
                    m.mergeLinkDataArray(_this.props.linkDataArray);
                }
                if (_this.props.modelData !== undefined) {
                    m.assignAllDataProperties(m.modelData, _this.props.modelData);
                }
            }, 'gojs-react init merge');
        });
    };
    ReactDiagram.prototype.componentWillUnmount = function () {
        var diagram = this.getDiagram();
        if (diagram !== null) {
            diagram.div = null;
            if (this.modelChangedListener !== null) {
                diagram.removeModelChangedListener(this.modelChangedListener);
                this.modelChangedListener = null;
            }
        }
    };
    ReactDiagram.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        if (nextProps.skipsDiagramUpdate)
            return false;
        if (nextProps.nodeDataArray === this.props.nodeDataArray &&
            nextProps.linkDataArray === this.props.linkDataArray &&
            nextProps.modelData === this.props.modelData)
            return false;
        return true;
    };
    ReactDiagram.prototype.componentDidUpdate = function (prevProps, prevState) {
        var diagram = this.getDiagram();
        if (diagram !== null) {
            var model = diagram.model;
            model.startTransaction('update data');
            model.mergeNodeDataArray(this.props.nodeDataArray);
            if (this.props.linkDataArray !== undefined && model instanceof GraphLinksModel) {
                model.mergeLinkDataArray(this.props.linkDataArray);
            }
            if (this.props.modelData !== undefined) {
                model.assignAllDataProperties(model.modelData, this.props.modelData);
            }
            model.commitTransaction('update data');
        }
    };
    ReactDiagram.prototype.render = function () {
        return (createElement("div", { ref: this.divRef, className: this.props.divClassName }));
    };
    return ReactDiagram;
}(Component));

var ReactOverview = (function (_super) {
    __extends(ReactOverview, _super);
    function ReactOverview(props) {
        var _this = _super.call(this, props) || this;
        _this.divRef = createRef();
        return _this;
    }
    ReactOverview.prototype.getOverview = function () {
        if (this.divRef.current === null)
            return null;
        return Diagram.fromDiv(this.divRef.current);
    };
    ReactOverview.prototype.componentDidMount = function () {
        if (this.divRef.current === null)
            return;
        var overview;
        if (this.props.initOverview !== undefined) {
            overview = this.props.initOverview();
        }
        else {
            overview = new Overview();
            overview.contentAlignment = Spot.Center;
        }
        overview.div = this.divRef.current;
        overview.observed = this.props.observedDiagram;
    };
    ReactOverview.prototype.componentWillUnmount = function () {
        var overview = this.getOverview();
        if (overview !== null) {
            overview.div = null;
            overview.observed = null;
        }
    };
    ReactOverview.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        if (nextProps.observedDiagram === this.props.observedDiagram)
            return false;
        return true;
    };
    ReactOverview.prototype.componentDidUpdate = function (prevProps, prevState) {
        var overview = this.getOverview();
        if (overview !== null) {
            overview.observed = this.props.observedDiagram;
        }
    };
    ReactOverview.prototype.render = function () {
        return (createElement("div", { ref: this.divRef, className: this.props.divClassName }));
    };
    return ReactOverview;
}(Component));

var ReactPalette = (function (_super) {
    __extends(ReactPalette, _super);
    function ReactPalette(props) {
        var _this = _super.call(this, props) || this;
        _this.divRef = createRef();
        return _this;
    }
    ReactPalette.prototype.getPalette = function () {
        if (this.divRef.current === null)
            return null;
        return Diagram.fromDiv(this.divRef.current);
    };
    ReactPalette.prototype.componentDidMount = function () {
        var _this = this;
        if (this.divRef.current === null)
            return;
        var palette = this.props.initPalette();
        palette.div = this.divRef.current;
        palette.delayInitialization(function () {
            var model = palette.model;
            model.commit(function (m) {
                m.mergeNodeDataArray(_this.props.nodeDataArray);
                if (_this.props.linkDataArray !== undefined && m instanceof GraphLinksModel) {
                    m.mergeLinkDataArray(_this.props.linkDataArray);
                }
                if (_this.props.modelData !== undefined) {
                    m.assignAllDataProperties(m.modelData, _this.props.modelData);
                }
            }, null);
        });
    };
    ReactPalette.prototype.componentWillUnmount = function () {
        var palette = this.getPalette();
        if (palette !== null) {
            palette.div = null;
        }
    };
    ReactPalette.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        if (nextProps.nodeDataArray === this.props.nodeDataArray &&
            nextProps.linkDataArray === this.props.linkDataArray &&
            nextProps.modelData === this.props.modelData)
            return false;
        return true;
    };
    ReactPalette.prototype.componentDidUpdate = function (prevProps, prevState) {
        var palette = this.getPalette();
        if (palette !== null) {
            var model = palette.model;
            model.startTransaction('update data');
            model.mergeNodeDataArray(this.props.nodeDataArray);
            if (this.props.linkDataArray !== undefined && model instanceof GraphLinksModel) {
                model.mergeLinkDataArray(this.props.linkDataArray);
            }
            if (this.props.modelData !== undefined) {
                model.assignAllDataProperties(model.modelData, this.props.modelData);
            }
            model.commitTransaction('update data');
        }
    };
    ReactPalette.prototype.render = function () {
        return (createElement("div", { ref: this.divRef, className: this.props.divClassName }));
    };
    return ReactPalette;
}(Component));

export { ReactDiagram, ReactOverview, ReactPalette };
